### DISCUSSION TITLE
Forbindelsen ADCData

### DISCUSSION THREAD

**Discussion by FoghThomsen** [Solvgraa-mager/E4PRJ4#52] [General]
Hej Adam (@AdamHoej-au)
Lige et par spørgsmål: 
1) Jeg taggede dig i et issue omkring krav til sensor, som du tog engang.. det er som om der er en masse forsvundet. Har du set det?
2) Forbindelsen ADCData er en parallelforbindelse (se nedestående billeder).. Det giver ikke mening at den er 8 forbindelser bred for du ender nok med at vælge en ADC med flere bit, gætter jeg på.. I så fald, med mere end 8 bit, giver det nok næppe mening at bruge en parallelforbindelse med eksempelvis 16 eller 24 forbindelser - derfor skal vi have lavet den forbindelse om. Det gør at jeg må ændre meget af mit analyseafsnit, hvorfor det kunne være fedt med en afklaring på nedestående, der gør at jeg kan begynde det arbejde der ligger i det.. 
2.1) Kan du lægge dig fast på et antal bit's opløsning? 
2.2) Hvilken slags forbindelse vil du gerne at ADCData er? SPI? (Du bestemmer, det er nemmeste at MCU'en bare tilpasses efter det når jeg alligevel skal lave analyse om).. 
3) ADCData protokol har jeg ladet stå åben, så du kan udfylde den med det du syntes giver mening :-)

![image](https://user-images.githubusercontent.com/61549345/114134672-69474a80-9908-11eb-8a64-3728936a2608.png)
![image](https://user-images.githubusercontent.com/61549345/114134709-76643980-9908-11eb-8f1f-625214900293.png)
![image](https://user-images.githubusercontent.com/61549345/114134736-7d8b4780-9908-11eb-979d-b5b5cbb7471c.png)

**Comment by AdamHoej-au** [comment]
Hej @AU636523 ! 

1. Jeg har set det - det er ikke mig der har rykket dem over i første omgang, og jeg må da indrømme jeg har undret mig - men ikke fået gjort mere ved det. Der må lige igangsættes en skattejagt på hvor de oprindelige krav befinder sig.
2. Jeg er pt. i gang med at prøve undersøge hvordan vi reelt set interfacer med 4-20mA og har været igang med at undersøge sensorere og ranges på disse - det arbejde der er lavet hidtil er at finde i journalens analyse afsnit.  Jeg kunne godt bruge en sparringspartner på det her - da ADC'er stadig er lidt.. specielle at arbejde med synes jeg. 
EDIT: Umiddelbart tænker jeg at SPI ville være optimalt - 4 wire interface og selvom der ikke er reele standarder må der være nogle defacto standarder der kan gøres brug af.
![image](https://user-images.githubusercontent.com/56192310/114137546-a7df0400-990c-11eb-93ac-8af4db994ce0.png)


3. Dette bliver forhåbentligt afklaret når der er mere styr på analysen af ADC I/O.

**Reply by FoghThomsen** [reply]
Hey @AdamHoej-au 

1. Det lyder ærgerligt. Jeg har ikke engang set dem efter din kategorisering og strukturering, så jeg ville ikke kunne genskabe det. Vil du tage opgaven på dig? 

2. 
2.1) Det vil være super, giv gerne besked når du ved hvor mange bit's opløsning du kommer til at arbejde med (det kommer til at give nogle ændringer hos mig nemlig).. 
2.2) OK, så bliver det jo bare SPI i stedet :-) Jeg laver om i mit afsnit så SPI kommer til at indgå fremfor den 8-ledede parallelle forbindelse.. Hvis du vil se på de krav, som nævnt i pkt. 1, så skal jeg nok tage denne del og rette i det fælles struktureringsafsnit.. Jeg syntes vi kan bruge SPI som standard for data-laget af kommunikationen og CMOS 3.3V-standard for logiske levels på det fysiske plan.. 

3.
Cool - det betyder, i modsætning til de to andre punkter, ikke så meget for mig for nuværende heller.. Jeg kan komme ret langt før jeg bliver afhængelig af nogle svar på det i mine afsnit 👍 

Er det hele cool med dig?

God weekend.. 

/Simon

**Comment by AdamHoej-au** [comment]
@AU636523

1.  Det lyder som ansvarsfralæggelse fra min side af - men jeg har ikke rørt de krav. Jeg var med til at snakke om dem, men jeg har ikke været igang med at fjerne / rykke sensorkrav - men jo jeg vil gerne kigge på dem. 
2. Hvis mine udregninger ikke er helt knaste (kig gerne i Overleaf(5.2.7) så rammer vi minimum 10 bit - den vil jeg gerne lige ta' en snak om!  
SPI er jeg desuden helt klar på at køre igennem med. Det burde være relativt ligetil at implementere noget fornuftigt!

Adam

**Reply by FoghThomsen** [reply]
@AdamHoej-au,
1) Hm, okay, men så betyder det vel bare at de aldrig blev lavet - for jeg lavede dem nemlig heller ikke, da det var dig der blev sat på issuet.. Men fedt - tror de vigtige, så fedt at du gider lave dem.. 
2) Antal bit: Cool! (fedt afsnit du har fået strikket sammen!) SPI: Fedt, så prøver jeg at arbejde videre med det.

**Reply by FoghThomsen** [reply]
@AdamHoej-au 

Har du i øvrigt overvejet hvordan vandstand skal blive lavet om til et SPI-signal? :) En idé jeg lige ville dele: MISO, MOSI, CLK og SS kunne hver repræsentere én vandstandssensor.. Så når multiplex-signalet signalerer at det vil aflæse vandstand, så sørger et stykke diskret logik for at de 4 forbindelser i stedet forbindes til hver deres vandstandssensor :)

**Reply by AdamHoej-au** [reply]
@AU636523 
Hej Simon, jeg sad lige og tjekkede databladet til din MCU, og ser at der er 10-bit ADC i den... Er der en grund til at vi ikke udnytter det, og så bare kobler sensorene direkte på med en smule prekonditionering? Jeg føler godt nok det er spild af resourcer at designe med multiplexer og kontrolbits, hvis det hele kan kobles direkte på én MCU og styres med dertilhørende kode.. Lad mig høre din mening!

**Reply by FoghThomsen** [reply]
Hej Adam @AdamHoej-au, 
Jeg kan godt forstå hvad du mener - og tanken har også strejfet mig på et tidspunkt. Jeg er praktisk talt er færdig med struktur, analyse, design (~20 sider og rigtig rigtig mange timers arbejde) for den opdeling vi oprindeligt aftalte - så fra mit perspektiv ligger de spildte ressourcer hos mig hvis vi ændrer det så sent i processen :-)

Egentligt var jeg af den opfattelse at du gerne ville arbejde med at lave hele ADC-delen, så det kommer nok også lidt bag på mig. 

Men jeg er selvfølgelig villig til at vi gør noget, hvis det giver bedst mening. 

Hvordan havde du forestillet dig den nye struktur skulle være?

**Reply by AdamHoej-au** [reply]
@AU636523  Hele sidste besked lød mere negativt fra min side end tiltænkt - beklager! 


Jeg blev blot overrasket over at se at der  var en ADC blok i den PIC du har valgt, og at det ikke var blevet nævnt, da der har været meget snak om optimering af enhederne. 
Det er ikke fordi jeg ikke vil arbejde med ADC-delen, men jeg vil også gerne have det mest optimale produkt ud af det, hvor vi forsvare vores komponenetvalg udfra de krav der er stillet.  

Den nye struktur ville være som den tidligere, blot med de to hovedblokke koblet til én SensorController blok - der designmæssigt består af en 16 bens PIC med RS-232 input/output til CC, og dertil 3 * ADC inputs  med 4-20mA måleinputs, samt et 2 digitale inputs til Vandstandsensorer samt forsyning.
En anden gang kunne det være vi skulle have kommunikeret lidt tydligere/tidligere om fremgangen i delblokken..

**Comment by AdamHoej-au** [comment]
Jeg har snakket med Bjarne og fået følgende oplysninger

- PR's temperatursensor (-80℃:400℃) kan indstille til et span ned til 10℃ (PR's kan indstilles vha. DIP-switches på transmitter).. Hvilket giver en helt ny analyse af rækkeviden for de forskellige sensorer.
- Lav kvalitetssensorer kan ikke indstilles!
- Metodikken vi har tænkt med Multiplexer og ADC vil fungere i praksis - men dette er ikke den "rigtige" måde at gøre det på. Men som det ser ud nu kan vi bare køre på med det!!

> Vi bruger PIC controllers med Delta Sigma modulatorere implementeret i operationsforstærkerer, hvorfor forbruget af en sensorenhed bliver holdt til et minimum.
